specs/design-black-box.md

```diff
@@ -72,10 +72,121 @@ specific ConfigMaps/Secrets.
 
 ## Design Details
 
-### Public API
-
-To be documented during POC implementation.
-
+The proposed design is based on the PoC that contains draft implementation and provides the base idea of a target design. PoC is available in
+[argo-cd#fargo/engine](https://github.com/alexmt/argo-cd/tree/fargo/engine) and [flux#fargo](https://github.com/alexmt/flux/tree/fargo) 
+
+The GitOps engine API consists of three main packages:
+* The `utils` package. It consist of loosely coupled packages that implements K8S resource diffing, health assessment, etc
+* The `app` package that Application data structure definition and CRUD client interface.
+* The `engine` package that leverages `utils` and uses provided set of Applications as a configuration source.
+
+```
+gitops-engine
+|-- pkg
+|   |-- utils
+|   |   |-- diff   # provides Kubernetes resource diffing functionality
+|   |   |-- kube   # provides utility methods to interact with Kubernetes
+|   |   |-- lue    # provides utility methods to run Lua scripts
+|   |   |-- apps   # provides utility methods to manipulate Applications (e.g. start sync operation, wait for sync operation, force reconciliation)
+|   |   `-- health # provides Kubernetes resources health assessment
+|   |-- app
+|   |   |-- apis       # contains data structures that describe Application
+|   |   `-- client     # contains client that provides Application CRUD operations
+|   `-- engine     # the engine implementation
+```
+
+The engine is responsible for reconciliation only which includes:
+- Interacting with the Kubernetes API: load and maintain the cluster state; pushing required changes to the Kubernetes API.
+- Reconciliation logic: based on provides target resource definition find which resources should be updated/deleted/created.
```

Comment:
I cannot parse this :S . typo?

Reply:
Looks like it was too early in the morning :) fixed.